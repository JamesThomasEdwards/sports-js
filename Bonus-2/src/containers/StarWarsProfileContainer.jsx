// React;
import React from 'react';
import { Component } from 'react';
// Files;
import Profile from '../components/Profile.jsx';

export default class StarWarsProfileContainer extends Component {
    state = {
        profiles: []
    };

    // gets number of pages;
    getPageCount = () => {
        return fetch(`https://swapi.co/api/people/?page=1`)
            .then(res => res.json())
            .then(pageOneObject => Math.ceil(pageOneObject.count / 10))
            .catch(error => console.log(error));
    };
    // takes in number of pages;
    // fetches each page and gets all profile data in an array;
    // then filters through it and sent to setttingState;
    getCharacterProfiles = (numberOfPages) => {
        const profiles = [];
        for (let i = 1; i <= numberOfPages; i++) {
            profiles.push(fetch(`https://swapi.co/api/people/?page=${i}`)
                .then(res => res.json())
                .catch(err => new Error(err)))
        };
        return Promise.all(profiles).then(values => {
            return values.reduce((accum, page) => {
                page.results.forEach(profile => {
                    accum.push({
                        name: profile.name,
                        birth_year: profile.birth_year,
                        height: profile.height,
                        mass: profile.mass
                    });
                });
                return accum
            }, []);
        });
    };
    // sets state with new filtered prfile data array;
    settingState = (filteredProfileData) => {
        this.setState(() => ({ profiles: filteredProfileData }));
    };

    allCharacterProfileData = async () => {
        let numOfPages = await this.getPageCount();
        let getAllProfiles = await this.getCharacterProfiles(numOfPages);
        let settingCorrectData = await this.settingState(getAllProfiles);
        return settingCorrectData;
    };
    componentDidMount() {
        this.allCharacterProfileData();
    };
    render() {
        const profiles = this.state.profiles.map((profile, ind) => {
            return <Profile
                id={ind}
                key={'Profile' + ind}
                profileData={profile} />
        });
        return (
            <div>
                <h1 className='profile-table-container__h1'>
                    Star Wars Character Profiles
                </h1>
                <div className='profile-table-container'>
                    {profiles}
                </div>
            </div>
        );
    };
};