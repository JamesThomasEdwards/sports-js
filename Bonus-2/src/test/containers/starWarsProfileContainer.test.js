import React from 'react';
import { shallow } from 'enzyme';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import StarWarsProfileContainer from '../../containers/StarWarsProfileContainer';

configure({ adapter: new Adapter() });

test('should render StarWarsProfileContainer component correctly', () => {
    const renderer = new ReactShallowRenderer();
    renderer.render(<StarWarsProfileContainer />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
});