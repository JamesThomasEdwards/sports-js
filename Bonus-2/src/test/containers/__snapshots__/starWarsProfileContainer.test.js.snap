// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render StarWarsProfileContainer component correctly 1`] = `
<div>
  <h1
    className="profile-table-container__h1"
  >
    Star Wars Character Profiles
  </h1>
  <div
    className="profile-table-container"
  />
</div>
`;
